packages/react-components/react-carousel-preview/library/src/components/CarouselCard/useCarouselCard.ts

```diff
@@ -14,43 +16,97 @@ import { useCarouselStore_unstable } from '../useCarouselStore';
  * @param props - props from this instance of CarouselCard
  * @param ref - reference to root HTMLDivElement of CarouselCard
  */
+
 export const useCarouselCard_unstable = (
   props: CarouselCardProps,
   ref: React.Ref<HTMLDivElement>,
 ): CarouselCardState => {
+  'use no memo';
+
+  const { circular } = useCarouselContext_unstable();
+
+  // We rely on render-heavy context in circular so we want to avoid if disabled
+  // As circular is static, this doesn't break rule of hooks
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  return circular ? useCarouselCardCircular(props, ref) : useCarouselCardCore(props, ref);
```

Comment:
Just playing devil's advocate here but I think this can be refactored to strictly follow the rules of hooks with negligible perf impact.

`useCarouselCardCircular()` calls `useCarouselCardCore()` so the latter is always called in either code path. We can refactor this so something like:

```ts
const state = useCarouselCardCore(props, ref);
return useCarouselCardCircular(props, ref, state);
```

`useCarouselCardCircular()` calls `useCarouselContext_unstable()` to get the `circular` value but we've already called this in the parent hook. If we pursue the strict refactor we should eliminate the call in `useCarouselCard_unstable()`.

For here it's just a matter of using the value of `circular` to avoid lots of the heavy computation work in `useCarouselCardCircular()`. It cannot all be avoided of course and this is where the refactor gets a bit messy and perhaps harder to read and maintain.

Reply:
Our main perf impact here isn't so much circular itself, but the hooks that are called afterwards to get the remaining context if circular is enabled - in circular, all cards will need to be re-rendered on index change as they reposition themselves, however, in non-circular, we only need to re-render the active and non-active cards (and even this can potentially be avoided, as we may add an intersection observer to each card to detect when it is on screen and should be aria-hidden or not), while the parent slider handles positioning.

Since we would likely need to break hooks rules to avoid those re-renders it felt cleaner to use this format, which was a pattern we appropriated from v9 tree, although I agree ideally the hooks would be consistent as breaking them feels a little risky.